core/src/main/java/com/linecorp/armeria/internal/common/stream/HttpDataDeframer.java

```diff
@@ -0,0 +1,417 @@
+         * Returns a newly retained slice of this {@link ByteBuf} at the current {@code readerIndex}
+         * and increases the {@code readerIndex} by the specified {@code length}.
+         *
+         * @throws IndexOutOfBoundsException if the specified {@code length} is greater than {@link #length()}
+         */
+        public ByteBuf readBytes(int length) {
+            // fast path
+            final ByteBuf firstBuf = queue.peek();
+            if (firstBuf.readableBytes() >= length) {
+                return firstBuf.readRetainedSlice(length);
+            }
+
+            // slow path
+            ByteBuf value = null;
+            int remained = length;
+            for (ByteBuf buf : queue) {
+                if (!buf.isReadable()) {
+                    continue;
+                }
+
+                final int readSize = Math.min(remained, buf.readableBytes());
+                if (readSize == length) {
+                    return buf.readRetainedSlice(readSize);
+                } else {
+                    if (value == null) {
+                        value = alloc.buffer(length);
+                    }
+                    value.writeBytes(buf, readSize);
+                    remained -= readSize;
+                    if (remained == 0) {
+                        break;
+                    }
+                }
+            }
+
+            if (remained > 0) {
+                ReferenceCountUtil.release(value);
+                throw new IndexOutOfBoundsException(
+                        "readerIndex: " + (readerIndex() - length + remained) +
+                        " (expected: 0 <= readerIndex + " + length + " <= length(" + length() + "))");
+            }
+
+            return value;
+        }
+
+        /**
+         * Returns a byte at the specified absolute {@code index} in this {@link HttpDataReader}.
+         * This method does not modify {@code readerIndex} of this {@link HttpDataReader}.
+         */
+        public byte getByte(int index) {
+            int remained = index;
+            for (ByteBuf buf : queue) {
+                final int length = buf.capacity();
+                if (length == 0) {
+                    continue;
+                } else if (length >= remained) {
+                    return buf.getByte(remained);
+                } else {
+                    remained -= length;
+                }
+            }
+
+            throw new IndexOutOfBoundsException(
+                    "index: " + index + " (expected: 0 <= index < length(" + length() + "))");
+        }
+
+        /**
+         * Discards the {@link ByteBuf}s that has been fully consumed and is not readable anymore.
+         */
+        private void discardReadBytes() {
+            for (;;) {
+                final ByteBuf buf = queue.peek();
+                if (buf != null && !buf.isReadable()) {
+                    queue.remove().release();
+                } else {
+                    break;
+                }
```

Comment:
Could use `queue.poll()` that returns `null` when it's empty.

Reply:
Let me use `queue.peek()` because the `buf` is only removed when it is not readable.